### DISCUSSION TITLE
traceidratio vs parentbased_traceidratio

### DISCUSSION THREAD

**Discussion by itsmaneesh** [open-telemetry/opentelemetry-java-instrumentation#6304] [Q&A]
Hi,
Can you please clarify the sampling behavior of `parentbased_traceidratio` compared to `traceidratio` sampler strategy options: (for OTEL_TRACES_SAMPLER env variable) with OpenTelemetry auto-instrumentation.

I have gone through couple of discussions threads and understood that `parentbased_traceidratio` strategy will sample the traces (as per OTEL_TRACES_SAMPLER_ARG value) at upstream level (typically at the time of trace Id creation) and ensure the completeness of the trace (meaning, all the expected spans will be associated with the trace)

Discussion threads:
https://github.com/open-telemetry/opentelemetry-collector-contrib/discussions/4462
https://github.com/open-telemetry/opentelemetry-java/discussions/3235

It would be great if you can point to me to any doc or clarify here the exact difference between `parentbased_traceidratio` and  `traceidratio`

Thank you,
Maneesh

**Comment by mateuszrzeszutek** [comment]
Hey @itsmaneesh 

You can think of `parentbased_` as a decorator over the sampler referenced after the underscore (that is, `traceidratio` in this case).
Take a look at the [official samplers doc](https://opentelemetry.io/docs/reference/specification/trace/sdk/#parentbased), in particular the table that describes the behavior of the sampler depending on the input.
If the span being sampled is the root span of the trace, `parentbased_traceidratio` will behave exactly the same as `traceidratio`; if it's not the root of the trace, then you can look up how it'll behave in the table.

**Comment by laurit** [comment]
You can view the parentbased one as sampling the whole trace, the traces you'll end up with are complete. The non parent based one will sample individual spans, so you could end up with a bunch of incomplete traces. Use the parentbased one unless you know what you are doing.

**Reply by itsmaneesh** [reply]
@laurit Thank you and understood that, `parentbased_traceidratio` can be used to achieve completeness of traces during sampling.
Also, can you clarify the sampling algorithm of `parentbased_traceidratio`? I couldn't clearly understood the algorithm from the below links:
https://opentelemetry.io/docs/reference/specification/trace/sdk/#requirements-for-traceidratiobased-sampler-algorithm
https://github.com/open-telemetry/opentelemetry-specification/issues/1413

For e.g.: 
With OTEL_TRACES_SAMPLER as `parentbased_traceidratio` and OTEL_TRACES_SAMPLER_ARG  as `0.1` strategy, Do we get to see 10% of traces being sampled during a particular time window (i.e.: 10 % of traces those got generated during 1 sec or any other time window like a Rate Limiting Fixed or Sliding Window algorithm??) ?? OR Otel will wait for 100 traces to be generated and sample 10% of those traces?

If service A (a first service in the workflow at which the trace and root span get created) is receiving 100 requests per sec then do we get to see just 10 traces out of those 100 requests /sec. (with OTEL_TRACES_SAMPLER_ARG  as `0.1`) and rest 90 will be dropped from sampling ?

**Reply by laurit** [reply]
You an examine the algorithm used in java implementation at https://github.com/open-telemetry/opentelemetry-java/blob/main/sdk/trace/src/main/java/io/opentelemetry/sdk/trace/samplers/TraceIdRatioBasedSampler.java
Basically it means that you specify the probability of trace getting sampled e.g. with OTEL_TRACES_SAMPLER_ARG as `0.1` you could say that given large enough time window it is likely that 10% of traces will be sampled.

**Reply by itsmaneesh** [reply]
@laurit I had a look at the class to understand the algorithm.
But, have couple of questions:
1. Algorithm seem to extract absolute long value out of the `traceid` hexadecimal (base16) string from offset 16 (mid of the string length) till end and compare that long value against the `(long) (ratio * Long.MAX_VALUE)` where `ratio` defined with env variable OTEL_TRACES_SAMPLER_ARG and 0.0 < `ratio` < `1.0`.  Question is how does this ensure 10% of traces will be sampled, provided the `0.1` is the ratio.?
2. From your statement:

> with OTEL_TRACES_SAMPLER_ARG as 0.1 you could say that **given large enough time window** it is likely that 10% of traces will be sampled

what is the "given large enough time window" ? Can you please clarify.

**Reply by laurit** [reply]
Trace id is 128 bits/16 bytes. My understanding of this is that given a uniformly distributed trace id its upper and lower half also have uniform distribution. So we can use only half of the trace id which is more convenient to deal with. To simplify the calculations we don't want to deal with negative numbers so we use Math.abs after which we are left with positive numbers and Long.MIN_VALUE as the only negative number. As there is no positive value that corresponds to Long.MIN_VALUE it is its own absolute value, we'll deal with this anomaly by ignoring it and making it so that for our purposes Long.MIN_VALUE is the same as zero. Now we have reduced trace id to a number that is in rang of 0 to Long.MAX_VALUE and the assumption is that it is still uniformly distributed. So if we want to get 10% of the traces we can compute 0.1*Long.MAX_VALUE and pick traces where the reduced trace id is below that value.

> what is the "given large enough time window" ? Can you please clarify.

You could think of this as an analogy of tossing a coin. Even if you start off with 5 straight heads eventually, after you have repeated the toss enough times, you'll have roughly the same amount of heads and tails.

**Reply by itsmaneesh** [reply]
@laurit  Will test this configuration with `OTEL_TRACES_SAMPLER_ARG as 0.1` and let you know if there is any deviation from our expectation of 10 % traces to be sampled. Thank you.

But, If we have two services ServiceA (upstream) and ServiceB (downstream) (both are using auto-instrumentation) have different OTEL_TRACES_SAMPLER_ARG values then how does the sampling strategy works:

For e.g.:
Scenaio#1: _Upstream system has lower value for OTEL_TRACES_SAMPLER_ARG 
ServiceA (upstream) has OTEL_TRACES_SAMPLER_ARG as 0.1 and ServiceB (downstream) has OTEL_TRACES_SAMPLER_ARG as 0.3._

Does the lower value of upstream system takes precedence and entire sampling will be based on value 0.1 i.e. 10 % traces sampled.?
Please confirm.

Scenario#2: _Downstream system has lower value for OTEL_TRACES_SAMPLER_ARG 
ServiceA(upstream) has OTEL_TRACES_SAMPLER_ARG  as 0.3 and ServiceB(downstream) has OTEL_TRACES_SAMPLER_ARG  as 0.1._

Does the upstream system value takes precedence despite being a higher value than the downstream system value? i.e.: 30 % traces will be sampled through out flow from ServiceA -> ServiceB. (downstream system's lower sampling value will be ignored as the upstream system's sampling value takes precedence? 
Please clarify.

**Reply by laurit** [reply]
If you are using parent based sampler then the sampling decision will be done by the service that creates the trace (ServiceA).
If you are not using parent based sampler then for scenario#1 the lower value from ServiceA should take precedence. This is because the algorithm used in java traceidratio sampler is deterministic. All traces that are sampled with 0.1 are also sampled with 0.3. I have no idea what would happen if you mix services written with java with services written in some other language. For scenario#2 the lower value of ServiceB will take effect and some traces will be missing spans from ServiceB.
At least this is how I understand it.

**Reply by trask** [reply]
@itsmaneesh also check out https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/sdk.md#requirements-for-traceidratiobased-sampler-algorithm
